failback.go

```diff
@@ -158,5 +159,8 @@ func (r *FailBack) startResetTimer() chan struct{} {
 
 // Returns true is the response is considered successful given the options.
 func (r *FailBack) isSuccessResponse(a *dns.Msg) bool {
-	return a == nil || !(r.opt.ServfailError && a.Rcode == dns.RcodeServerFailure)
+	aLen := len(a.Answer)
+	return a == nil || !(r.opt.ServfailError && a.Rcode == dns.RcodeServerFailure) &&
```

Comment:
This is very hard to read and it looks wrong as well, should probably be `a == nil || !(r.opt.ServfailError && a.Rcode == dns.RcodeServerFailure) || ...`. It might be best to rewrite it with simple if statements and avoid the boolean algebra here:

```go
if a == nil {
  return true
}
if !(r.opt.ServfailError && a.Rcode == dns.RcodeServerFailure) {
  return true
}
if r.opt.EmptyError {
...
}
```

Reply:
No, the logic was correct and I think the code is clever because there's no redundancy (the rest of the condition only runs when it hasn't hit bad case). The new code is double the lines for no real benefit other than that it's maybe a tiny bit faster to understand.